zend_abstract_interface/methods/methods.h

```diff
@@ -0,0 +1,48 @@
+#ifndef ZAI_METHODS_H
+#define ZAI_METHODS_H
+
+#include <main/php.h>
+#include <stdbool.h>
+
+/* Work in progress
+ *
+ * The long-term goal is to provide ZAI data-structure shims so that the
+ * following APIs are the same across all PHP versions. For now we will provide
+ * ZAI seams using native Zend Engine data structures.
+ */
+
+/* Looks up a class entry directly from the CG(class_table). The lookup is
+ * case-sensitive so the class name should be lowercase. The class name should
+ * not contain a root-scope '\' prefix. Does not invoke the autoloader. Returns
+ * NULL if the class entry could not be found. Because the class lookup occurs
+ * from the compiler globals, this can be called safely outside of a request
+ * context.
+ */
+zend_class_entry *zai_class_lookup_ex(const char *cname, size_t cname_len TSRMLS_DC);
+
+/* Calls a method on an instance of 'object' without passing any arguments.
+ * Caller must pass in a NULL pointer-pointer for the return value 'retval'.
+ * Caller must dtor a non-NULL 'retval' after the call:
+ *
+ *   if (retval) {
+ *     zval_ptr_dtor(&retval);
+ *   }
+ *
+ * Methods cannot be called outside of a request context so this MUST be called
+ * from within a request context (after RINIT and before RSHUTDOWN). A crash
+ * will occur if this is called outside of a request context.
+ */
+bool zai_call_method_without_args_ex(zval *object, const char *method, size_t method_len, zval **retval TSRMLS_DC);
```

Comment:
Can we properly detect request context and return false if we're not inside one? Then we can run tests since they won't crash. If this is too expensive then we can consider moving it to debug-only or something.

Reply:
Good idea. I've aded this check in e346b91. I believe this only applies to static method calls since a crash will occur when trying to instantiate an object outside of a request context. So I updated the formally crashy test to a static method call test.